### DISCUSSION TITLE
Select All options in a Form

### DISCUSSION THREAD

**Discussion by jonathanTack** [cortexrd/Knack-Toolkit-Library#333] [Q&A] [answered]
Looking for a way to select all options in a Knack Form. Have a Form with a Multiple-select Field (connection to parent object). want to be able to select ALL checkboxes. 

![image](https://github.com/user-attachments/assets/64f3ca63-d0e8-4f41-8c5d-41813d95f00f)

thank you!

**Comment by cortexrd** [comment]
Hi @jonathanTack,

Nice to see you here!!  A warm welcome to you :)

I had something brewing for a while that does just that.

I'd call it "**Set Form Default Values**".  Keyword would be **_sfdv**.

Let me check if I can do that now...

Norm

**Reply by jonathanTack** [reply]
Thanks Norm! 
I watched your and Carl’s videos for hours last night, getting oriented and learning all I need about KTL. 
The ask here is for a toggle “select all options” and to undo, ie “select none”.

thanks very much ! 
Going to impress client if this is possible .

**Comment by chprojectsupport** [comment]
Hi Johnathan,Great to see you here and really pleased to learn that you’ve been watching our videos. They can be a great cure for insomnia. 😂Knowing Normand he won’t rest until he has a solution. He has been completely amazing and helped me overcome a complex date/time calculation field with a new keyword.Hope you’re keeping safe and well. Still flying I presume ✈️🌎Carl HolmesOn 31 Jul 2024, at 15:04, jonathanTack ***@***.***> wrote:﻿
Thanks Norm!
I watched your and Carl’s videos for hours last night, getting oriented and learning all I need about KTL.
The ask here is for a toggle “select all options” and to undo, ie “select none”.
thanks very much !
Going to impress client if this is possible .

**Reply by jonathanTack** [reply]
hey Carl! thanks for noticing and chiming in! yes sir, still have that privilege of heights, thanks for asking! took me awhile to respond to ya here because I had a trip for a couple hours and had to finish that and then get settled at the hotel here in Indianapolis, Indiana. Your videos are really well done. The KTL seems to be a great asset that is worth the investment to learn. Keep the videos coming!

**Comment by cortexrd** [comment]
@jonathanTack,

I have something for you to try.  Currently, it only works with Many to Many connected fields with the Checkbox format.  I think that's what you need for now.

Here's the keyword to be added in your form's description (or title):

`_sfdv=[fieldLabelOrId, value1, value2, ...]`

You can use the field label as shown in your view, or its ID.
For the values, you can either add them individually, or you can use the special parameter **ktlAll** to select them all, like this:

`_sfdv=[fieldLabelOrId, ktlAll]`

Give it a try in Dev mode and let me know if it works.

Cheers,
Norm

**Comment by cortexrd** [comment]
To switch to Dev or other modes, see how this is done here:

[Debugging and Switching Special KTL versions](https://github.com/cortexrd/Knack-Toolkit-Library/discussions/186)

**Comment by cortexrd** [comment]
I've added support for these field types:

1. All text fields
2. Simple selection connection dropdowns
3. Checkboxes
4. Radio buttons
5. Date/Time

You can use multiple groups like this to support many fields in a single form:

`_sfdv=[[Short Text, Text Sample], [Currency, 123.45], [Date/Time, 07/15/2024, 1:23pm], [Multiple Choice Radio buttons, Second Choice], [Yes/No Radio Buttons, Yes], [Multiple Choice Checkboxes, First Choice, Third Choice], [One to many dropdown, 0023590716839300]`


Norm

**Reply by jonathanTack** [reply]
@cortexrd thanks very much for driving this effort! I know i'm asking a lot being my very first step into KTL... when this is solved, i need to buy you and Carl a coffee at the least. 
i have implemented the KTL Configurator, added Developer User Role to myself, and switched the KTL code to 'Dev'. I then applied the keyword to my title as follows:  _sfdv=[field_189]. 
At this point, nothing is showing up that is different. 

I could take a screenshare and send it to you - maybe you can detect something else i need to adjust. do you have a certain email address i can send a link to (don't want the public to see the app). Appreciate you!

**Reply by cortexrd** [reply]
You must specify a default value at least.

Examples:
`_sfdv=[field_189, ktlAll]`
to select all checkboxes

`_sfdv=[field_189, First Choice, Second Choice, Fourth Choice]`
to select only those 3 choices.

**Reply by cortexrd** [reply]
BTW guys... it a good time to rename that keyword if you don't like it.

Any other suggestions?

_def (Defaults)
_fdv (Form Default Values)
_sdv (Set Default Values)

Norm

**Reply by jonathanTack** [reply]
no change still ... i also tried to use the field name

![image](https://github.com/user-attachments/assets/78a3a18c-656a-45f8-88f2-e0359ca4c632)

**Reply by cortexrd** [reply]
Oh... you changed from checkboxes to dropdown.

Dropdown don't support multiple selections yet.

You can try these options:

1. if you want to use a dropdown, replace ktlAll by an entry that exists in the list.  Must be an exact match (all text with spaces), but not case sensitive.
2. go back to checkboxes and use the ktlAll as second param.

ND

**Reply by jonathanTack** [reply]
so that screenshot is in the Builder... in the app , they are checkboxes. now that you mention it, that is odd, for the Builder to have a dropdown but the App has checkboxes. 
i checked the field properties; field_189 is authentically checkboxes.

![image](https://github.com/user-attachments/assets/44cb89b4-7fe1-4042-a09c-4af3173e2821)

**Reply by jonathanTack** [reply]
could it be something with the settings in the Configurator? I set all the params in the configurator the same as Carl demonstrated in his [video ](https://www.youtube.com/watch?v=2_vVcIhdRlo)

**Reply by cortexrd** [reply]
No, it's not related to the KTL Configurator.
I have the same setup as yours and here it works ok.

Can we have a Zoom call to see what's happening live?
Or give me access to your page (or similar).

**Reply by jonathanTack** [reply]
can do zoom rigth away ... what's your email? or email me avdataco@gmail.com

**Comment by chprojectsupport** [comment]
Following this thread, be interested if you got it working JT. Carl HolmesOn 1 Aug 2024, at 00:43, jonathanTack ***@***.***> wrote:﻿
can do zoom rigth away ... what's your email? or email me ***@***.***

**Reply by jonathanTack** [reply]
Carl - Norm spent some quality time with me last night to get it working. and he added some extra widgets as well, which are much appreciated. all you say about him is true : ) happy to have both of you on my side.

**Comment by cortexrd** [comment]
Guys, I have decided to rename the keyword from _sfdv to **_sfv (Set Field Value)**

Reasons: 

1. I just can't remember that darn keyword!!!  _sfv is easier.
2. It will eventually support more than just forms, like grids and searches, with API calls.
3. It opens the door to using ktlRefVal, where we'll be able to fetch values from another view/field and use it in this view.

I will do an official release in a few minutes, so change it in your app please.

Norm

**Reply by jonathanTack** [reply]
ok i'm back now... had an exciting day and then a surprise short trip down to Kentucky. 
When i got on my app, the select all was working, and then i switched the keyword to _sfv=[Communities, ktlAll], and now it's not selecting all anymore. i'll give it more time to adjust. 
but besides that - Normand, thanks for all your investment yesterday. I'm gonna jam on this project starting now and let you know how it progresses.

**Reply by jonathanTack** [reply]
on your point 3 - i've been doing this with a JavaScript, and which has been working well. One key to my current setup is that my Client wanted to have form presets that are Logged-On User dependent. i did achieve this, but it took much effort. I told my Client that if he wants to add another user, that the Dev time will take awhile. So as you're working on ktlRefVal, this could be good for your awareness.

**Comment by cortexrd** [comment]
Jon,

Remove the Dev code please.  This could be the reason why it's not working anymore.  Usually, Dev code is short lived since I use it all the time for various tests.

Go back to Prod mode (letter "p").

Thanks for the nice comments, it's appreciated :)

**Reply by jonathanTack** [reply]
done, and good!

**Comment by jonathanTack** [comment, marked as answer]
I believe I can count this answered now! Thanks @cortexrd  for your extensive and prompt help on this!

Here's what I did: 
**1) The main form opens and presents the current selections, with no KTL keywords implemented. There are 2 links on this scene, namely "Select All" and "Select None". These are the toggles so that the user can decide here to make ad-hoc selections or All or None.** 
![image](https://github.com/user-attachments/assets/70f61ba5-f7ce-497e-beb5-d6a8c7c17a6d)

**2) If "Select All" is clicked, then the next Scene opens which looks identical, but all selections are ticked. Here, I use the keyword _sfv=[Communities,ktlAll]. The link is also now "Revert Selection", which returns the navigation to the original Current Selections scene.**
![image](https://github.com/user-attachments/assets/d637ba6a-0322-4c48-9296-bbde4f4ad993)

**3)  If "Select None " is clicked, then the next Scene opens which looks identical, but all selections are blank. Here, I use the keyword _sfv=[Communities,]. The link is also now "Revert Selection", which returns the navigation to the original Current Selections scene.** 
![image](https://github.com/user-attachments/assets/8a993311-1cf7-48ce-80c8-f2e5b7dca8e3)

This is grand - thank you so much!
